### PR TITLE
[TR] PIM-4238: revamp family import

### PR DESCRIPTION
**Pull request by solivier**
| Q                    | A
| -------------------- | ---
| Bug fix?             | No
| New feature?         | No
| BC breaks?           | yes
| CI currently passes? | Yes
| Tests pass?          | Yes
| Scenarios pass?      | Yes
| Checkstyle issues?*  | No
| PMD issues?**        | No
| Changelog updated?   | Yes
| Fixed tickets        | PIM-4238
| DB schema updated?   | No
| Migration script?    | No
| Doc PR               | No

CE: http://ci.akeneo.com/job/pim-pr-behat/2721/console

### CODE REVIEW

File: spec/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderSpec.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            ->shouldBeCalled();
+
+        $family = $this->createFamily(true);
+        $family->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Model\FamilyInterface');
+        $family->getAttributes()->shouldHaveCount(1);
+        $family->getAttributes()->first()->shouldBeEqualTo($identifierAttribute);
+        $family->getAttributeRequirements()->shouldHaveCount(2);
+        $family->getAttributeRequirements()->shouldBeEqualTo([
+            'anyCode_print' => $printRequirement,
+            'anyCode_ecommerce' => $ecommerceRequirement
```
Comment (skeleton) [inline spec/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderSpec.php]: please align

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilder.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            }
+        }
+    }
+
+    /**
+     * @return \Pim\Bundle\CatalogBundle\Entity\Channel[]
+     */
+    protected function getChannels()
+    {
+        return $this->channelRepository->findBy([]);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilder.php]: findAll() ?

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Pim\Bundle\CatalogBundle\Builder;
+
+use Pim\Bundle\CatalogBundle\Model\FamilyInterface;
+
+/**
+ * Product builder interface
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2014 Akeneo SAS (http://www.akeneo.com)
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php]: 2015

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public function createFamily($withIdentifier = false);
+
+    /**
+     * @param FamilyInterface $family
+     * @param array           $data
+     *
+     * @return null
+     */
+    public function setAttributeRequirements(FamilyInterface $family, array $data);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php]: please add @throws \InvalidArgumentException

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public function setLabels(FamilyInterface $family, array $data);
+
+    /**
+     * @param FamilyInterface $family
+     * @param array           $data
+     *
+     * @return null
+     */
+    public function addAttributes(FamilyInterface $family, array $data);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php]: idem

File: src/Pim/Bundle/CatalogBundle/Doctrine/Common/Saver/FamilySaver.php [hunk 6]
```diff
@@ -65,4 +66,27 @@ public function save($family, array $options = [])
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function saveAll(array $families, array $options = [])
+    {
+        if (empty($families)) {
+            return;
+        }
+
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/Common/Saver/FamilySaver.php]: remove extra blank line

File: src/Pim/Bundle/CatalogBundle/Updater/FamilyUpdater.php [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+    /** @var IdentifiableObjectRepositoryInterface */
+    protected $familyRepository;
+
+    /** @var FamilyBuilderInterface */
+    protected $familyBuilder;
+
+    /**
+     * @param IdentifiableObjectRepositoryInterface $familyRepository
+     * @param FamilyBuilderInterface $familyBuilder
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Updater/FamilyUpdater.php]: please align

File: src/Pim/Bundle/InstallerBundle/Resources/config/batch_jobs.yml [hunk 8]
```diff
@@ -116,21 +116,21 @@ connector:
...
 
-        fixtures_family_yml:
-            title: pim_installer.jobs.fixtures_family_yml.title
+        fixtures_family_csv:
+            title: pim_installer.jobs.fixtures_family_csv.title
             type:  fixtures
             steps:
                 validation:
-                    title: pim_base_connector.jobs.fixtures_family_yml.validation.title
+                    title: pim_base_connector.jobs.fixtures_family_csv.validation.title
```
Comment (skeleton) [inline src/Pim/Bundle/InstallerBundle/Resources/config/batch_jobs.yml]: Change "pim_base_connector" to "pim_connector" for title too

File: src/Pim/Component/Connector/ArrayConverter/Flat/FamilyStandardConverter.php [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * TODO: to improve to make it more understandable
+     *
+     * @param array  $convertedItem
+     * @param string $field
+     * @param mixed  $data
+     *
+     * @return array
+     */
+    protected function convertField($convertedItem, $field, $data)
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/FamilyStandardConverter.php]: type hint

File: src/Pim/Component/Connector/Processor/Denormalization/FamilyProcessor.php [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * @param StandardArrayConverterInterface       $familyConverter
+     * @param IdentifiableObjectRepositoryInterface $repository
+     * @param FamilyBuilderInterface                $familyBuilder
+     * @param ObjectUpdaterInterface                $updater
+     * @param ValidatorInterface                    $validator
+     */
+    public function __construct(
+        StandardArrayConverterInterface $familyConverter,
```
Comment (skeleton) [inline src/Pim/Component/Connector/Processor/Denormalization/FamilyProcessor.php]: please set $repository as the first param in the construct

File: src/Pim/Component/Connector/Processor/Denormalization/FamilyProcessor.php [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        if (null === $family) {
+            return $this->familyBuilder->createFamily();
+        }
+
+        return $family;
+    }
+
+    /**
+     * @param FamilyInterface $family
+     * @param array                    $convertedItem
```
Comment (skeleton) [inline src/Pim/Component/Connector/Processor/Denormalization/FamilyProcessor.php]: please align

**Comment by nidup** [comment]
changelog for FamilyFactory transformed to FamilyBuilder

File: src/Pim/Bundle/TransformBundle/Transformer/FamilyTransformer.php [hunk 12]
```diff
@@ -59,7 +55,7 @@ public function __construct(
      */
     protected function createEntity($class, array $data)
     {
-        return $this->factory->createFamily();
+        return $this->familyBuilder->createFamily();
```
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Transformer/FamilyTransformer.php]: $this->familyBuilder->createFamily(true); to be sure that the requirement for the sku is added, that was the case before you changed the factory/family

File: CHANGELOG.md [hunk 13]
```diff
@@ -142,6 +142,7 @@
 - Constructor of `Pim\Bundle\EnrichBundle\Manager\SequentialEditManager` has been changed
 - Depreciate and change constructor of `Pim\Bundle\TransformBundle\Builder\FieldNameBuilder`
 - Replace the argument ProductManager by ProductRepositoryInterface in the constructor of `Pim\Bundle\CatalogBundle\Validator\Constraints\UniqueVariantAxisValidator`
+- Transform `Pim\Bundle\CatalogBundle\Factory\FamilyFactory` in `Pim\Bundle\CatalogBundle\Builder\FamilyBuilder`
```
Comment (phaseinducer) [inline CHANGELOG.md]: rather use "transform something **into** something else"  ;)

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilder.php [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+            $family->setAttributeAsLabel($attribute);
+        } else {
+            throw new \InvalidArgumentException(
+                sprintf('Attribute with "%s" code does not exist', $data)
+            );
+        }
+    }
+
+    /**
+     * @return \Pim\Bundle\CatalogBundle\Entity\Channel[]
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilder.php]: Import class please, and `ChannelInterface[]` would be better also.

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php [hunk 15]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Pim\Bundle\CatalogBundle\Builder;
+
+use Pim\Bundle\CatalogBundle\Model\FamilyInterface;
+
+/**
+ * Product builder interface
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php]: c/p typo

File: src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public function createFamily($withIdentifier = false);
+
+    /**
+     * @param FamilyInterface $family
+     * @param array           $data
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return null
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Builder/FamilyBuilderInterface.php]: Do we explicitly add a `@return` even when it's `null` ?

File: src/Pim/Component/Catalog/Updater/FamilyUpdater.php [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    /**
+     * @param FamilyInterface $family
+     * @param string          $field
+     * @param mixed           $data
+     */
+    protected function setData(FamilyInterface $family, $field, $data)
+    {
+        if ('labels' === $field) {
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Updater/FamilyUpdater.php]: IMHO a switch is more readable

File: src/Pim/Component/Connector/ArrayConverter/Flat/FamilyStandardConverter.php [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+            $convertedItem['requirements'][$requirementsLocale] = explode(',', $data);
+        } else {
+            switch ($field) {
+                case 'code':
+                    $convertedItem[$field] = (string) $data;
+                    break;
+                case 'attributes':
+                    $convertedItem[$field] = explode(',', $data);
+                    break;
+                case 'attribute_as_label':
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/FamilyStandardConverter.php]: "attribute_as_label" & "code" are identical.

File: src/Pim/Component/Connector/Processor/Denormalization/FamilyProcessor.php [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+     * @param ValidatorInterface                    $validator
+     */
+    public function __construct(
+        IdentifiableObjectRepositoryInterface $repository,
+        StandardArrayConverterInterface $familyConverter,
+        FamilyFactory $familyFactory,
+        ObjectUpdaterInterface $updater,
+        ValidatorInterface $validator
+    ) {
+        parent::__construct($repository);
```
Comment (skeleton) [inline src/Pim/Component/Connector/Processor/Denormalization/FamilyProcessor.php]: add a blank line after parent::__construct($repository);

**Pull request merged**